src/ast/passes/semantic_analyser.cpp

```diff
@@ -3462,25 +3463,41 @@ void SemanticAnalyser::visit(Program &program)
 
 int SemanticAnalyser::analyse()
 {
-  // Multiple passes to handle variables being used before they are defined
   std::string errors;
 
-  int num_passes = listing_ ? 1 : num_passes_;
-  for (pass_ = 1; pass_ <= num_passes; pass_++) {
+  int last_num_unresolved = 0;
+  pass_ = 1;
+  // Multiple passes to handle variables being used before they are defined
+  while (true) {
+    num_unresolved_ = 0;
+
     Visit(ctx_.root);
+
     errors = err_.str();
     if (!errors.empty()) {
       out_ << errors;
       return pass_;
     }
-  }
 
-  return 0;
+    if (is_final_pass_) {
+      return 0;
+    }
+
+    if (num_unresolved_ > 0 &&
+        (last_num_unresolved == 0 || num_unresolved_ < last_num_unresolved)) {
+      // If we're making progress, keep making passes
+      last_num_unresolved = num_unresolved_;
+    } else {
+      is_final_pass_ = true;
+    }
+
+    pass_++;
```

Comment:
Maybe `LOG(BUG)` if pass_ goes above 1000 or something? In case there's a resolution but it'd be better to fail hard than spin forever. Failing louder, esp. in automated environments is probably better here.

Reply:
Actually, I think the progress checking functionality should handle this case. Even if we start with a HUGE number of unresolved types, it needs to make progress or else it exists with an error.

Reply:
Ah, yeah. Progress checks should be catching it.